src/python/pants/backend/python/dependency_inference/module_mapper.py

```diff
@@ -84,6 +98,28 @@ def find_all_python_projects(all_targets: AllTargets) -> AllPythonTargets:
     return AllPythonTargets(tuple(first_party), tuple(third_party))
 
 
+# -----------------------------------------------------------------------------------------------
+# Namespace packages collection
+# -----------------------------------------------------------------------------------------------
+
+
+class PythonNamespacePackages(frozenset):
+    """A set of Python namespace packages."""
+
+
+@rule(desc="Creating a collection of namespace-packages", level=LogLevel.DEBUG)
+async def collect_namespace_packages(
+    all_python_targets: AllPythonTargets,
+    python_setup: PythonSetup,
+) -> PythonNamespacePackages:
+    return PythonNamespacePackages(
+        module.split(".")[0]
+        for tgt in chain(all_python_targets.first_party, all_python_targets.third_party)
+        for module in (tgt.get(PythonRequirementModulesField).value or [])
+        if "." in module
```

Comment:
A "namespace package" is a very specific thing in Python (see https://peps.python.org/pep-0420/#namespace-packages-today) etc. And this isn't checking for any of that. I think it's assuming that the parent of any module with a `.` in it is a namespace package, but that need not be the case.

And anyway, I don't think you actually need to know about namespace packages at all. I think there's a simpler solution here, similar to what I outlined in the [discussion](https://github.com/pantsbuild/pants/discussions/17286) but even more straightforward.

Basically, instead of the new enum, we can track the distance of the parent (0=we found the symbol itself, 1=we found its parent, etc.)

Then we resolve ambiguities by taking the provider with the smallest distance. And only if there are multiple of those do we warn and skip the dep.

Does that make sense?

Reply:
In `map_module_to_address`, there is `request.module` (which is `"pacifica.cli"`), while the provided results are `ModuleProvider`s with `Address`es (`Address(//:root#pacifica-cli), Address(pacifica_db/pacifica/__init__.py:../pacifica-db-src-sources)`).

So IIUC, I need to make both the 1st and 3rd party `providers_for_module` functions either return `AddressAndDistancePairs` or `AddressAndModulePairs` and then do the resolution in `map_module_to_address`.
Is that correct?

Another option is to (somehow) get the mapping from [the root `BUILD` file](https://github.com/TalAmuyal/pants-repro/blob/namespace_collision_example/BUILD) which has the exact mapping we want:
```python
{
    "module_mapping": {
        provider.addr.generated_name: [request.module],
    }
}
```
I've pushed a third commit that serves as an example for that one.

Reply:
Yeah, I was thinking that we add an `ancestry` field of type `int | None` to `ModuleProvider`. We set it to None when we generate the mappings, and then return a copy, with `ancestry` set appropriately, in `FirstPartyPythonModuleMapping._providers_for_resolve` and in `ThirdPartyPythonModuleMapping._providers_for_resolve`. Then we use that field to fix ambiguities in `map_module_to_address`.

Or - since having that field be None in the mapping is a bit hacky, we could also create a different type to represent the "provider + ancestry": `PotentialModuleProvider` or something.

Reply:
BTW Looking at the existing code, I don't like the fact that `{First,Third}PartyPythonModuleMapping` extends a `FrozenDict` instead of encapsulating one. We don't actually use it as a dict anywhere, so this is just being too clever for no reason.  But that's not for this change...

Reply:
OK, so I gave it another go.
I've started with making the `PotentialModuleProvider` and then removed it as I noticed `request.module` is available in `map_module_to_address`, where I assume we should have used the `.ancestry` field in the first place.
Also, I'm not sure that `calc_ancestry` has the logic you intended, but it does work both for the repro and our internal repo.

Please LMK what you think.

Reply:
Oh, yeah, this is actually neat! And doesn't require adding fields or data structures. I like the idea, will take a closer look at the implementation now.

Reply:
See my comment, this was a neat idea, but unfortunately `calc_ancestry` is not correct in general.